DigitalLearningSolutions.Web.Tests/ControllerHelpers/ControllerContextHelper.cs

```diff
@@ -0,0 +1,56 @@
+﻿namespace DigitalLearningSolutions.Web.Tests.ControllerHelpers
+{
+    using System;
+    using System.Security.Claims;
+    using System.Threading.Tasks;
+    using DigitalLearningSolutions.Data.Tests.Helpers;
+    using FakeItEasy;
+    using Microsoft.AspNetCore.Authentication;
+    using Microsoft.AspNetCore.Http;
+    using Microsoft.AspNetCore.Mvc;
+    using Microsoft.AspNetCore.Mvc.Routing;
+    using Microsoft.AspNetCore.Mvc.ViewFeatures;
+
+    public static class ControllerContextHelper
+    {
+        public static void SetUpControllerWithUser<T>(ref T controller, string authenticationType = "")
```

Comment:
A few questions about this class:
- What's the role of authentication type being a parameter rather than something hardcoded?
- Do you think that instead of passing the ref around and adding a controller as a param, we write extension methods similar to CustomClaimHelper that will allow us to use syntax like controller.SetUpMockUser()? This way we can also make the syntax a bit cleaner (I think you should be able to use something like SetUpControllerWithUser(this Controller controller, ...) { } + use controller.SetUpThis; controller.SetUpThat; in the tests)
- The "Set up controller with user" is actually setting up a mock user session and a fake temp data, which seems independent of the user session. Could we potentially separate this out into two methods that can be called where relevant?
- The "SetUpControllerWithServices" is actually setting up the user session too behind the scenes which could be confusing to those that call it - could we make these two independent. Maybe something like a following set of methods:
- SetUserSession
- SetUpFakeTempData (although this is so small and rare that we could just set it up in the test setup and actually have reference to the fake - then we can check whether particular temp data is attempted to be saved/cleared etc in the tests).
- SetUpFakeAuthenticationService

Reply:
The authentication type determines what type of user is created:
- if it is an empty string then an unauthenticated user is set up (if this wasn't set up, then controller checks such as User.Identity.IsAuthenticated to redirect would fail with a null user)
- if it is anything other than an empty string, an authenticated user is set up.
I left it as a parameter so the same code can be run to set things up with just that one parameter deciding if a user should be authenticated, but I suppose a single bool UserIsAuthenticated would help make the method signature easier to understand.

The other remarks make sense, I'll look into them. Might have to be careful with the order the methods are called in as one will set up new ControllerContext and new DefaultHttpContext while the others will just modify them, but it should be feasible.

Reply:
I've implemented those changes. The session setting has been removed as it wasn't being used anywhere. I also created the SetUpDefaultController for the initial set up of the default http context which will always need to be called before the others so I'm not sure if there is a better way around it.

Reply:
I think that's fine! What I noticed is that you set these methods up more like a builder (returning itself), which is quite nice. What do you think about calling these methods "WithMockUser" etc so when you create the controller it reads like this:

```
var controller = new XXXController()
.WithDefaultContext()
.WithMockUser()
.WithMockServices();
```

I'm happy with either, but I think either way it would make sense to put the .SetUpXXX or .WithXXX onto new lines when defining the controller (for readability). Otherwise, looks great!

Reply:
Yeah I put the return in so that it could all be done on a single line when the controller is first defined. Seeing as that is how it is usually used, I think the "WithMockX" make more sense. I'll make those changes.